queries/appSettings.ts

```diff
@@ -1,81 +1,61 @@
 import 'server-only';
+import { type z } from 'zod';
+import { createAppSetting } from '~/actions/appSettings';
 import { UNCONFIGURED_TIMEOUT } from '~/fresco.config';
 import { createCachedFunction } from '~/lib/cache';
+import { type appSettingSchema } from '~/schemas/appSettings';
 import { prisma } from '~/utils/db';
 
-type AppSettingsKeys =
-  | 'configured'
-  | 'allowAnonymousRecruitment'
-  | 'limitInterviews'
-  | 'initializedAt'
-  | 'installationId';
-
-type AppSettings = Record<AppSettingsKeys, string>;
-
-const DEFAULT_SETTINGS: Record<AppSettingsKeys, string> = {
-  configured: 'false',
-  allowAnonymousRecruitment: 'false',
-  limitInterviews: 'true',
-  initializedAt: new Date().toISOString(),
-  installationId: createId(),
-};
+// Generic function to get an app setting
+export async function getAppSetting<
+  Key extends keyof z.infer<typeof appSettingSchema>,
+>(key: Key): Promise<z.infer<typeof appSettingSchema>[Key]> {
+  const initializedAt = await prisma.appSettings.findUnique({
+    where: { key: 'initializedAt' },
+  });
 
-const calculateIsExpired = (configured: boolean, initializedAt: Date) =>
-  !configured && initializedAt.getTime() < Date.now() - UNCONFIGURED_TIMEOUT;
+  if (!initializedAt) {
+    // initialize the app
+    // todo: should these come from fresco.config?
+    await createAppSetting('initializedAt', new Date());
+    await createAppSetting('configured', false);
+    await createAppSetting('allowAnonymousRecruitment', false);
+    await createAppSetting('limitInterviews', true);
+    await createAppSetting('installationId', createId());
+  }
 
-const getAppSettings = createCachedFunction(async () => {
-  const appSettingsRecords = await prisma.appSettings.findMany({
-    select: { key: true, value: true },
+  const keyValue = await prisma.appSettings.findUnique({
+    where: { key },
   });
 
-  // Check if there are any existing settings
-  if (appSettingsRecords.length === 0) {
-    // No app settings exist, so create all default settings
-    await prisma.appSettings.createMany({
-      data: Object.entries(DEFAULT_SETTINGS).map(([key, value]) => ({
-        key: key as AppSettingsKeys,
-        value,
-      })),
-    });
-
-    return {
-      ...DEFAULT_SETTINGS,
-      configured: DEFAULT_SETTINGS.configured === 'true',
-      allowAnonymousRecruitment:
-        DEFAULT_SETTINGS.allowAnonymousRecruitment === 'true',
-      limitInterviews: DEFAULT_SETTINGS.limitInterviews === 'true',
-      initializedAt: new Date(DEFAULT_SETTINGS.initializedAt),
-      expired: calculateIsExpired(
-        DEFAULT_SETTINGS.configured === 'true',
-        new Date(DEFAULT_SETTINGS.initializedAt),
-      ),
-    };
+  if (!keyValue) {
+    return null;
   }
 
-  const appSettings = appSettingsRecords.reduce((acc, { key, value }) => {
-    acc[key as AppSettingsKeys] = value;
-    return acc;
-  }, {} as AppSettings);
+  // Parse the value based on the expected type
+  const parsedValue =
```

Comment:
I would expect you to be using either `preprocess()` or `coerce()` here. Zod normally wants to parse then validate, but because we already know the values coming from the db will be the wrong type for anything but strings, so you need to use those methods.

Maybe you should try writing some validation tests to make sure you get this working correctly?

Reply:
implemented using `preprocess()` and added validation tests [24c2a27](https://github.com/complexdatacollective/Fresco/pull/161/commits/24c2a27b6246ff76bd611535579698ad77b7f324)